rustls-mbedcrypto-provider/src/sign.rs

```diff
@@ -125,47 +138,47 @@ const RSA_SIGNATURE_SCHEME_PREFER_LIST: &[SignatureScheme] = &[
-        match self.pk_type {
-            mbedtls::pk::Type::Rsa | mbedtls::pk::Type::RsaAlt | mbedtls::pk::Type::RsassaPss => {
-                // choose a rsa schema
-                for scheme in RSA_SIGNATURE_SCHEME_PREFER_LIST {
-                    if offered.contains(scheme) {
-                        let signer = MbedTlsSigner(Arc::clone(&self.pk), *scheme);
-                        return Some(Box::new(signer));
-                    }
-                }
-                None
-            }
-            mbedtls::pk::Type::Eckey | mbedtls::pk::Type::EckeyDh | mbedtls::pk::Type::Ecdsa => {
-                let scheme = self
-                    .ec_signature_scheme
-                    .expect("validated");
-                if offered.contains(&scheme) {
-                    let signer = MbedTlsSigner(Arc::clone(&self.pk), scheme);
-                    return Some(Box::new(signer));
-                }
-                None
-            }
-            _ => None,
-        }
+        let scheme = get_signature_schema_from_offered(
+            self.pk_type,
+            offered,
+            self.ec_signature_scheme,
+            self.rsa_scheme_prefer_order_list,
+        )?;
+        let signer = MbedTlsSigner(Arc::clone(&self.pk), scheme);
+        Some(Box::new(signer))
     }
 
     fn algorithm(&self) -> rustls::SignatureAlgorithm {
         self.signature_algorithm
     }
 }
 
-fn pk_type_to_signature_algo(pk_type: mbedtls::pk::Type) -> rustls::SignatureAlgorithm {
-    use rustls::SignatureAlgorithm;
+/// Helper function to choose proper [`SignatureScheme`] based on given inputs
+pub fn get_signature_schema_from_offered(
+    pk_type: mbedtls::pk::Type,
+    offered: &[SignatureScheme],
+    ec_signature_scheme: Option<SignatureScheme>,
+    rsa_scheme_prefer_order_list: &[SignatureScheme],
+) -> Option<SignatureScheme> {
     match pk_type {
-        mbedtls::pk::Type::Rsa => SignatureAlgorithm::RSA,
-        mbedtls::pk::Type::Ecdsa => SignatureAlgorithm::ECDSA,
-        mbedtls::pk::Type::RsassaPss => SignatureAlgorithm::RSA,
-        mbedtls::pk::Type::RsaAlt => SignatureAlgorithm::RSA,
-        mbedtls::pk::Type::Eckey => SignatureAlgorithm::ECDSA,
-        mbedtls::pk::Type::EckeyDh => SignatureAlgorithm::Unknown(255),
-        mbedtls::pk::Type::Custom => SignatureAlgorithm::Unknown(255),
-        mbedtls::pk::Type::None => SignatureAlgorithm::Unknown(255),
+        mbedtls::pk::Type::Rsa | mbedtls::pk::Type::RsaAlt | mbedtls::pk::Type::RsassaPss => {
+            // choose a rsa schema
+            for scheme in rsa_scheme_prefer_order_list {
+                if offered.contains(scheme) {
+                    return Some(*scheme);
+                }
+            }
+            None
+        }
+        mbedtls::pk::Type::Eckey | mbedtls::pk::Type::EckeyDh | mbedtls::pk::Type::Ecdsa => {
+            let scheme = ec_signature_scheme.expect("validated");
```

Comment:
If we need to panic here, could we have a better error message?

Reply:
updated